### ISSUE TITLE
grid-box-sizing-001 fails due to missing file

### ISSUE THREAD

**Issue by frivoal** [w3c/web-platform-tests#7969]
The grid-box-sizing-001 file added by commit https://github.com/w3c/web-platform-tests/commit/d8f789878b359c58ebad395442ec4eb42050cbd9 fails on a 404 in all browsers due to a missing file.
`<script src="../support/check-layout-th.js"></script>`
does refer to a valid file that is in the source tree, but somehow once you load the test case from testharness, it fails: http://test.csswg.org/suites/css-ui-3_dev/nightly-unstable/html/grid-box-sizing-001.htm

I' not 100% sure what's going on, but maybe it's got something to do with the build system and relative paths?

**Comment by mrego** [comment]
Mmmm, ok I should have used `/support/check-layout-th.js` instead (like Flexbox does).
We've the same problem in all the grid tests:
http://test.csswg.org/suites/css-grid-1_dev/nightly-unstable/html/grid-positioned-items-implicit-grid-001.htm
I'll fix that, thanks @frivoal for reporting.

**Comment by mrego** [comment]
Sorry for the wrong comment, for Flexbox test they use `support/check-layout-th.js` as relative path that seems to be working fine:
* http://w3c-test.org/css/css-flexbox-1/percentage-heights-001.html
* http://w3c-test.org/css/css-flexbox-1/position-absolute-001.html
* http://w3c-test.org/css/css-flexbox-1/position-absolute-002.html
* http://w3c-test.org/css/css-flexbox-1/position-absolute-003.html
* http://w3c-test.org/css/css-flexbox-1/position-absolute-004.html

The same for tables:
* http://w3c-test.org/css/css-tables-3/fixed-layout-calc-width-001.html
* http://w3c-test.org/css/css-tables-3/fixed-layout-excess-width-distribution-001.html

But in Grid Layout we've it under `css/css-grid-1/support/check-layout-th.js` and we use it from subfolders inside the test suite like `css/css-grid-1/abspos/`.
That seems to be working fine in w3c-test.org, but not in the other site:
* http://w3c-test.org/css/css-grid-1/abspos/grid-positioned-items-implicit-grid-001.html

What kind of path should I use?

As a related note, it's not nice at all that we have to keep 3 copies of the very same `check-layout-th.js` file on 3 test suites. More in the future will use it (a lot of layout tests in Blink and WebKit repos rely on this script), so it'd be awesome to find a solution and avoid the files duplication.

**Comment by gsnedders** [comment]
> I' not 100% sure what's going on, but maybe it's got something to do with the build system and relative paths?

Yup, the build system doesn't deal with relative paths. It copies the contents of the support directory adjacent to the test, and nothing else.

> As a related note, it's not nice at all that we have to keep 3 copies of the very same check-layout-th.js file on 3 test suites. More in the future will use it (a lot of layout tests in Blink and WebKit repos rely on this script), so it'd be awesome to find a solution and avoid the files duplication.

We have two options: change the build system to find all support files a test includes and move them around (note this has been proven impossible for tests that use JavaScript [HALTINGPROBLEM]), or get rid of the build system (and when I looked at it before, moving the CSS test harness away from relying on the build system was going to be hard).

**Comment by mrego** [comment]
> Yup, the build system doesn't deal with relative paths. It copies the contents of the support directory adjacent to the test, and nothing else.

Any suggestion about how to fix this specific case?
Do we need N `check-layout-th.js` in each Grid Layout subfolder (like `css/css-grid-1/abspos/support/check-layout-th.js`? Or flatten (remove subfolders) the Grid Layout test suite (which doesn't sound like a great idea either).

**Comment by gsnedders** [comment]
> Do we need N `check-layout-th.js` in each Grid Layout subfolder (like `css/css-grid-1/abspos/support/check-layout-th.js`?

I'd say we should do this in the short term. 😞

**Comment by mrego** [comment]
I'll wait for more feedback before doing something like that, as we're not going to be able to maintain all these copies synchronised in the long run.

CC @foolip as he asked about the duplicated `check-layout-th.js` files in a Blink review.

**Comment by plinss** [comment]
The build system currently moves all the files in a test suite into a single folder and copies the contents of the support folders into a single support folder. It's not currently smart enough to update the relative paths for support files when it does this (because in addition to having to deal with the src and href attributes it would also need to parse all the css and update relative paths there as well). While doing this was part of the long-term plan, it's not likely to happen as we're trying to migrate away from built test suites so it's not reasonable to invest a lot of time in the build system (and no-one has the time either).

The historical way of handling this has always been to use relative paths like 'support/<filename>' and duplicate the necessary files in 'support' folders for each directory that needs them in the source tree. (The build system does validate that the duplicated files are identical, but no-one is really monitoring its error output these days either.) If the number of duplicates is too much of a support cost (understandable if the files are changing), then putting all the test files that rely on the support file into a single directory is your next best bet.

The current plan is to modify the build system/test harness to use the source files directly (or at least maintain the same directory structure in the output), so when that's done, then relative paths anywhere in the source repo can be used. No ETA when that will happen as I'm currently full time on other projects and not many others care to dive into the build/harness code.

**Comment by plinss** [comment]
(FWIW, the build _does_ update the relative paths of reference files, since it only has to look at link tags, so reference files can be anywhere in the source repo and don't need to be duplicated. There's just a lot more places and ways support files get referenced.)

**Comment by mrego** [comment]
Thanks @plinss for the detailed explanation.

I'm wondering if it could be possible to move `check-layout-th.js` into `web-platform-tests/resources/` folder. That way we could use it like we do with `testharness.js`, just with:
```html
 <script src="/resources/testharness.js"></script>
 <script src="/resources/testharnessreport.js"></script>
 <script src="/resources/check-layout-th.js"></script>
```

That way we'd avoid to have that file duplicated in several CSS Test Suites (currently 3 but probably more in the future if more Blink suites are exported into WPT).
But I don't know if it'd be acceptable to have such a file under `/resources`.
WDYT?

**Comment by plinss** [comment]
I'm not the maintainer/gatekeeper for the resources folder, not sure who is these days, @gsnedders may know. If it's a file commonly used in different test suites (and is generic functionality), then that seems like a reasonable solution to me.

**Comment by mrego** [comment]
Ok, so this should be fixed now by #7993.

**Comment by RByers** [comment]
Confirmed it [appears to be running fine now](https://wpt.fyi/css/css-grid/grid-model).
